gobblin-modules/google-ingestion/src/main/java/gobblin/ingestion/google/AsyncIteratorWithDataSink.java

```diff
@@ -0,0 +1,83 @@
+package gobblin.ingestion.google;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+
+import lombok.extern.slf4j.Slf4j;
+
+
+@Slf4j
+public abstract class AsyncIteratorWithDataSink<T> implements Iterator<T> {
+  private final Object lock = new Object();
+  private volatile Throwable exceptionInProducerThread = null;
+  private Thread _producerThread;
+  protected LinkedBlockingDeque<T> _dataSink = new LinkedBlockingDeque<>(2000);
+
+  @Override
+  public boolean hasNext() {
+    initialize();
+    if (!_dataSink.isEmpty()) {
+      return true;
+    }
+    try {
+      T next = _dataSink.poll(1, TimeUnit.SECONDS);
+      while (next == null) {
+        if (_producerThread.isAlive()) {
+          //Job not done yet. Keep waiting...
+          next = _dataSink.poll(1, TimeUnit.SECONDS);
+        } else {
+          synchronized (lock) {
+            if (exceptionInProducerThread != null) {
+              throw new RuntimeException(
+                  String.format("Found exception in producer thread %s", _producerThread.getName()),
+                  exceptionInProducerThread);
+            }
+          }
+          log.info("Producer job has finished. No more query data in the queue.");
+          return false;
+        }
+      }
+      //Must put it back. Implement in this way because LinkedBlockingDeque doesn't support blocking peek.
+      _dataSink.putFirst(next);
```

Comment:
No, it should be putFirst here, because I am trying to mimic the behavior of doing a blocking peek. At line 26, I am doing a _dataSink.poll, which is polling the first element from the queue. If something is there in the queue, I put it back as the first element, and it will keep the original order.